core/src/main/java/org/infinispan/commands/remote/ClusteredGetCommand.java

```diff
@@ -78,24 +79,23 @@ public void initialize(InvocationContextFactory icf, CommandsFactory commandsFac
    public CompletableFuture<Object> invokeAsync() throws Throwable {
       // make sure the get command doesn't perform a remote call
       // as our caller is already calling the ClusteredGetCommand on all the relevant nodes
-      long flagBitSet = EnumUtil.bitSetOf(Flag.SKIP_REMOTE_LOOKUP, Flag.CACHE_MODE_LOCAL);
+      // CACHE_MODE_LOCAL is not used as it can be used when we want to ignore the ownership with respect to reads
+      long flagBitSet = EnumUtil.bitSetOf(Flag.SKIP_REMOTE_LOOKUP);
       GetCacheEntryCommand command = commandsFactory.buildGetCacheEntryCommand(key, EnumUtil.mergeBitSets(flagBitSet, getFlagsBitSet()));
+      command.setTopologyId(topologyId);
       InvocationContext invocationContext = icf.createRemoteInvocationContextForCommand(command, getOrigin());
       CompletableFuture<Object> future = invoker.invokeAsync(invocationContext, command);
       return future.thenApply(rv -> {
-         CacheEntry cacheEntry = (CacheEntry) rv;
-         if (cacheEntry == null) {
-            if (trace) log.trace("Did not find anything, returning null");
-            return null;
-         }
          //this might happen if the value was fetched from a cache loader
-         if (cacheEntry instanceof MVCCEntry) {
+         if (rv instanceof MVCCEntry) {
             if (trace) log.trace("Handling an internal cache entry...");
-            MVCCEntry mvccEntry = (MVCCEntry) cacheEntry;
+            MVCCEntry mvccEntry = (MVCCEntry) rv;
             return entryFactory.createValue(mvccEntry);
-         } else {
-            InternalCacheEntry internalCacheEntry = (InternalCacheEntry) cacheEntry;
+         } else if (rv instanceof InternalCacheEntry) {
+            InternalCacheEntry internalCacheEntry = (InternalCacheEntry) rv;
             return internalCacheEntry.toInternalCacheValue();
+         } else {
+            return rv;
```

Comment:
Since commands can return aribitrary value and sometimes you need to return failure (though different than an exception - e.g. including value), the interceptors have to anticipate this. Or, we could really use just exceptions, though I am not in favour of that - one of the changes I haven't squeezed into the PR is replacing `WriteCommand.updateStatusFromRemoteResponse` with `UnsuccessfulResponse` carrying value.